### DISCUSSION TITLE
Imported events in remote calendars should appear in the proper spot for the user's local time

### DISCUSSION THREAD

**Discussion by dcoales** [davish/obsidian-full-calendar#358] [Feature Requests]
I have an event in my Office 365 Calendar that was created by a colleague in America.  The event is scheduled for 8.30am their time but 12:30 my time.  Outlook shows the event in my diary correctly at 12:30 but when I import this via the ics import it shows up in Full Calendar at 8:30 in the morning.

Is there something I can do to fix this or is it an issue with the ics import ?

**Comment by davish** [comment]
Timezones are notoriously tricky and right now, I'm assuming that the time given in the ics import is the time that an event should be displayed. I'll need to dig deeper into the ICS spec to pull out timezone information, and allow users to set their local timezone. I'll put this in the queue of bugs to work on!

**Comment by davish** [comment]
Adding a "help wanted" label to this issue because timezones are notoriously difficult, but if anyone is interested in working on this problem, it would be greatly appreciated!

**Comment by greenwar** [comment]
+1 for a google calendar on macos. timezones def tricky af.

**Comment by dcoales** [comment]
Thanks very much

**Comment by FabianoLothor** [comment]
In the meanwhile, this manual change at `obsidian-full-calendar/main.js` is doing the job for my `.ics` files:

```js
function expandICalEvents2(iCalExpander, range) {
  let rangeStart = addDays(range.start, -1);
  let rangeEnd = addDays(range.end, 1);
  let iCalRes = iCalExpander.between(rangeStart, rangeEnd);
  let expanded = [];
  for (let iCalEvent of iCalRes.events) {
    expanded.push(__spreadProps(__spreadValues({}, buildNonDateProps2(iCalEvent)), {
//    start: iCalEvent.startDate.toString(),
//      end: specifiesEnd2(iCalEvent) && iCalEvent.endDate ? iCalEvent.endDate.toString() : null
start: iCalEvent.startDate.timezone ? moment(new Date(new Date(iCalEvent.startDate.toString()).getTime() + (((new Date().setMilliseconds(0) - Date.UTC.apply(null, new Date().toLocaleString("ja", { timeZone: Intl.supportedValuesOf('timeZone').indexOf(iCalEvent.startDate.timezone) > -1 ? iCalEvent.startDate.timezone: Intl.DateTimeFormat().resolvedOptions().timeZone }).split(/[/\s:]/).map((p, i) => i === 1 ? 0 : p))) / 60 / 1000 - new Date().getTimezoneOffset()) * 60 * 1000 ))).format("YYYY-MM-DDTHH:mm:ss") : iCalEvent.startDate.toString(),
end: specifiesEnd2(iCalEvent) && iCalEvent.endDate ? ( iCalEvent.endDate.timezone ? moment(new Date(new Date(iCalEvent.endDate.toString()).getTime() + (((new Date().setMilliseconds(0) - Date.UTC.apply(null, new Date().toLocaleString("ja", { timeZone: Intl.supportedValuesOf('timeZone').indexOf(iCalEvent.endDate.timezone) > -1 ? iCalEvent.endDate.timezone : Intl.DateTimeFormat().resolvedOptions().timeZone }).split(/[/\s:]/).map((p, i) => i === 1 ? 0 : p))) / 60 / 1000 - new Date().getTimezoneOffset()) * 60 * 1000 )).toString()).format("YYYY-MM-DDTHH:mm:ss") : iCalEvent.endDate.toString()) : null
    }));
  }
  for (let iCalOccurence of iCalRes.occurrences) {
    let iCalEvent = iCalOccurence.item;
    expanded.push(__spreadProps(__spreadValues({}, buildNonDateProps2(iCalEvent)), {
//      start: iCalOccurence.startDate.toString(),
//      end: specifiesEnd2(iCalEvent) && iCalOccurence.endDate ? iCalOccurence.endDate.toString() : null
start: iCalEvent.startDate.timezone ? moment(new Date(new Date(iCalOccurence.startDate.toString()).getTime() + (((new Date().setMilliseconds(0) - Date.UTC.apply(null, new Date().toLocaleString("ja", { timeZone: Intl.supportedValuesOf('timeZone').indexOf(iCalEvent.startDate.timezone) > -1 ? iCalEvent.startDate.timezone : Intl.DateTimeFormat().resolvedOptions().timeZone }).split(/[/\s:]/).map((p, i) => i === 1 ? 0 : p))) / 60 / 1000 - new Date().getTimezoneOffset()) * 60 * 1000 ))).format("YYYY-MM-DDTHH:mm:ss") : iCalOccurence.startDate.toString(),
end: specifiesEnd2(iCalEvent) && iCalOccurence.endDate ? ( iCalEvent.endDate.timezone ? moment(new Date(new Date(iCalOccurence.endDate.toString()).getTime() + (((new Date().setMilliseconds(0) - Date.UTC.apply(null, new Date().toLocaleString("ja", { timeZone: Intl.supportedValuesOf('timeZone').indexOf(iCalEvent.endDate.timezone) > -1 ? iCalEvent.endDate.timezone : Intl.DateTimeFormat().resolvedOptions().timeZone }).split(/[/\s:]/).map((p, i) => i === 1 ? 0 : p))) / 60 / 1000 - new Date().getTimezoneOffset()) * 60 * 1000 )).toString()).format("YYYY-MM-DDTHH:mm:ss") : iCalOccurence.endDate.toString()) : null
    }));
  }
  return expanded;
}
```

Previous code:

```js
function expandICalEvents2(iCalExpander, range) {
  let rangeStart = addDays(range.start, -1);
  let rangeEnd = addDays(range.end, 1);
  let iCalRes = iCalExpander.between(rangeStart, rangeEnd);
  let expanded = [];
  for (let iCalEvent of iCalRes.events) {
    expanded.push(__spreadProps(__spreadValues({}, buildNonDateProps2(iCalEvent)), {
    start: iCalEvent.startDate.toString(),
      end: specifiesEnd2(iCalEvent) && iCalEvent.endDate ? iCalEvent.endDate.toString() : null
    }));
  }
  for (let iCalOccurence of iCalRes.occurrences) {
    let iCalEvent = iCalOccurence.item;
    expanded.push(__spreadProps(__spreadValues({}, buildNonDateProps2(iCalEvent)), {
      start: iCalOccurence.startDate.toString(),
      end: specifiesEnd2(iCalEvent) && iCalOccurence.endDate ? iCalOccurence.endDate.toString() : null
    }));
  }
  return expanded;
}
```

I know it isn't DRY, but it's easier this way to update the file using a bash+sed command.

---

The code won't work for all use cases, invalid ISO Timezones will be ignored and the default timezone will take its place.

Also, the code assumes that `moment` will be available, I tried to avoid it, but I didn't find a way to get `new Date().toJSON()` using its timezone, it always returns using `UTC`.

**Comment by vlitvin** [comment]
@FabianoLothor didn't work out for me. Do you happen to have any tips on troubleshooting this thing? It's a real deal breaker for me, cause I have team members in 5 different timezones (from -11 to +3 hours from me) and calendar records are all in the wrong places.

**Comment by pcause** [comment]
thnaks for this.  It messes up all day events.  Shows the event as during the day and busy and not in the all day header only.

**Comment by FabianoLothor** [comment]
Hey guys, sorry to hear about that, just revert the code and your `ics` events will be in the right place again.

@vlitvin it's hard to say my friend.

Basically what the code is doing is passing the right `startDate` and `endDate` based on the object provided by the `ics` file.

Everything will depend on how the `ics` file is returning the things.

If you want to dig, add a breakpoint or a `console.log` after the lines below:

```
here:
  for (let iCalEvent of iCalRes.events) {
    console.log(iCalEvent.startDate.toString(), iCalEvent.startDate.timezone, iCalEvent.endDate.toString(), iCalEvent.endDate.timezone)

and here:
  for (let iCalOccurence of iCalRes.occurrences) {
    let iCalEvent = iCalOccurence.item;
    console.log(iCalOccurence.startDate.toString(), iCalEvent.startDate.timezone, iCalOccurence.endDate.toString(), iCalEvent.endDate.timezone)
```
Those are the data that I'm using to convert the `toString` to my timezone.

It's tricky, indeed, because sometimes the `ics` doesn't provide the right unix-timestamp,

I'll update the code in the end of this week to something more clear and DRY to make it more understandable.

**Comment by vlitvin** [comment]
@FabianoLothor thanks a lot, I'm new to obsidian plugin development and npm in general. Though I have an understanding of the base logic behind that.
I'll try to dig into this later this week. Looks like I need to organize a testing stand for this, as experimenting in my main vault does bring more trouble than profit.

**Comment by CodyBontecou** [comment]
@FabianoLothor  How did you apply that code change? Did you fork the repo and load this script locally?
